File: java/game/src/main/java/example/GameManager.java
package example;

import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.ArrayList;

// This class showcases the use of Java Streams for filtering and sorting characters.
// It also demonstrates the use of Optional for null safety and modern Java features like method references.
// Additionally, it highlights the use of Comparator for custom sorting logic.
public class GameManager {

    private final List<Character> characters;

    public GameManager() {
        this.characters = new ArrayList<>();
    }

    public void addCharacter(Character c) {
        characters.add(c);
    }

    public void printCharacterByName(String name) {

        System.out.println("Output of: printCharacterByName()");

        Optional<Character> character = characters.stream()
                .filter(c -> c.getName().equals(name))
                .findFirst();

        character.ifPresentOrElse(
                c -> System.out.println("Character found: " + c.getName()),
                () -> System.out.println("Character not found.")
        );
    }

    public void printAliveCharacters() {

        System.out.println("Output of: printAliveCharacters()");

        // Adding a stream example to filter and sort characters
        List<Character> aliveCharacters = characters.stream()
                .filter(Character::isAlive)
                .sorted(Comparator.comparingInt(Character::getHealth).reversed())
                .toList();

        aliveCharacters.forEach(c -> System.out.println(c.getName() + " has " + c.getHealth() + " HP."));
    }

}

File: csharp/Game/GameManager.cs
namespace Game
{

    public class GameManager
    {
        private readonly List<Character> Characters = [];

        public void AddCharacter(Character c)
        {
            Characters.Add(c);
        }

        // In C#, LINQ is used for filtering and sorting, which is similar to Java Streams but has a different syntax.
        public void PrintAliveCharacters()
        {
            Console.WriteLine("Output of: PrintAliveCharacters()");
            var aliveCharacters = Characters
                .Where(c => c.IsAlive())
                .OrderByDescending(c => c.Health)
                .ToList();

            aliveCharacters.ForEach(c => Console.WriteLine($"{c.Name} has {c.Health} HP."));
        }

        // Adding a nullable example to handle null safety (C# equivalent to Java Optional)
        // The 'FirstOrDefault' method is used instead of Java's Optional to handle null values.
        public void PrintCharacterByName(string name)
        {
            Console.WriteLine("Output of: PrintCharacterByName()");
            var character = Characters.FirstOrDefault(c => string.Equals(c.Name, name, StringComparison.OrdinalIgnoreCase));

            Console.WriteLine(character != null
            ? $"Character found: {character.Name}"
            : "Character not found.");
        }

    }
}